{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b149134a-2254-4157-a285-59d54dd09e07",
   "metadata": {},
   "source": [
    "## Задание 1\n",
    "\n",
    "#### вопрос\n",
    "Выберите любой домен из базы данных Pfam. Единственное условие –\n",
    "\n",
    "в выравнивании seed должно быть хотя бы 20 последовательностей.\n",
    "\n",
    "Кратко опишите функцию домена. Заполните таблицу\n",
    "#### ответ\n",
    "\n",
    "note: accession в stockholm файле указан вместе с `.26`, хотя в браузере выводится без него\n",
    "\n",
    "\n",
    "|     AC Pfam|ID Pfam|#SEED|#All|#SW|#architectures|#3D|Taxonomy|#eukaryota|#archaea|#bacteria|\n",
    "|------------|-------|-----|----|---|--------------|---|--------|----------|--------|---------|\n",
    "|PF01625(.26)|   PMSR|  518| 66k|327|           232| 30|     42k|      7771|    2050|    54948|\n",
    "\n",
    "## Задание 2\n",
    "\n",
    "#### вопрос\n",
    "Скачайте выравнивание seed для Вашего домена и откройте его в\n",
    "JalView\n",
    "\n",
    "Раскрасьте выравнивание по свойствам аминокислот и консервативности\n",
    "\n",
    "колонок. Найдите в выравнивании консервативные и функционально\n",
    "\n",
    "консервативные колонки. Найдите хотя бы один максимально\n",
    "\n",
    "достоверный блок (см. презентацию к семинару 11). Заполните таблицу\n",
    "#### ответ\n",
    "\n",
    "|seed|100%-cons|   func-cons|max-reliable|\n",
    "|----|---------|------------|------------|\n",
    "| 518|  234:[H]|123:[IALMV] |     8-10   |\n",
    "|    |    8:[C]|            |            |\n",
    "|    |   10:[W]|            |            |\n",
    "|    |   46:[G]|            |            |\n",
    "\n",
    "\n",
    "Макимальный блок должен начинаться и заканчиваться\n",
    "\n",
    "консервативной или функционально-консервативной группой.\n",
    "\n",
    "Но в подобранном примере нету блока, длинее одного столбца,\n",
    "\n",
    "чтобы ни в одной колонке хотя бы одного нарушения условия функциональности.\n",
    "\n",
    "Выбрал этот блок, так как там нарушений минимальное кол-во - по 1-4 их 518.\n",
    "\n",
    "Считаю, что для (в основном) бактерий такую скидочку можно сделать\n",
    "\n",
    "## Задание 3\n",
    "\n",
    "#### вопрос\n",
    "Сравнение программ множественного выравнивания\n",
    "\n",
    "Выровняйте последовательности из seed любыми двумя программами\n",
    "\n",
    "множественного выравнивания.\n",
    "\n",
    "Сравните полученные выравнивания с выравниванием seed. Для этого\n",
    "\n",
    "необходимо найти совпадающие блоки в двух выравниваниях.\n",
    "\n",
    "#### ответ\n",
    "\n",
    "программа для сравнений:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9a51bcc8-02c6-4ae3-9173-18500b4a1ce8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ascii_zero = 48\n",
    "ascii_last = 126\n",
    "utf_shift = 128\n",
    "\n",
    "\n",
    "def score_chr(score) -> str:\n",
    "    \"\"\"\n",
    "    Maps columns score to single character (stockholm annotations allows only single char)\n",
    "    :return: printable ascii or utf-8\n",
    "    :raises ValueError: if unexpected type, got non-single-char-str, got negative score or bigger than printable utf-8\n",
    "    \"\"\"\n",
    "    if score is True:\n",
    "        return '1'\n",
    "    if score is False or score is None:\n",
    "        return '-'\n",
    "    if isinstance(score, str):\n",
    "        assert len(score) == 1\n",
    "        return score\n",
    "    if isinstance(score, int) or isinstance(score, float):\n",
    "        assert 0 <= score <= 0x10ffff\n",
    "        score = int(score)\n",
    "        # add code of ascii zero\n",
    "        # to print ascii digits and letters\n",
    "        if score + ascii_zero < ascii_last:\n",
    "            score += ascii_zero\n",
    "        else:\n",
    "            score += utf_shift\n",
    "        return chr(score)\n",
    "\n",
    "    raise ValueError(f'unsupported type for stockholm score annotation: {score}')\n",
    "\n",
    "\n",
    "aa_classes = {\n",
    "    'A': 'F',\n",
    "    'V': 'F',\n",
    "    'L': 'F',\n",
    "    'I': 'F',\n",
    "    'J': 'F',\n",
    "    'M': 'F',\n",
    "    'P': 'F',\n",
    "    'F': 'R',\n",
    "    'Y': 'R',\n",
    "    'W': 'R',\n",
    "    'G': 'Z',\n",
    "    'S': 'Z',\n",
    "    'T': 'Z',\n",
    "    'C': 'Z',\n",
    "    'Q': 'Z',\n",
    "    'N': 'Z',\n",
    "    'E': 'N',\n",
    "    'D': 'N',\n",
    "    'K': 'P',\n",
    "    'R': 'P',\n",
    "    'H': 'P',\n",
    "}\n",
    "\n",
    "\n",
    "for residue, cls in list(aa_classes.items()):\n",
    "    aa_classes[residue.lower()] = cls\n",
    "\n",
    "\n",
    "def are_residues_eq(x, y) -> bool:\n",
    "    \"\"\"\n",
    "    :return: True if X and Y equals\n",
    "    \"\"\"\n",
    "    return x in aa_classes and x.upper() == y.upper()\n",
    "\n",
    "\n",
    "def are_residues_similar(x, y) -> bool:\n",
    "    \"\"\"\n",
    "    :return: True if X and Y have similar properties\n",
    "    \"\"\"\n",
    "    x = aa_classes.get(x)\n",
    "    y = aa_classes.get(y)\n",
    "    return x is not None and x == y\n",
    "\n",
    "\n",
    "class SubsMatCmp:\n",
    "    \"\"\"\n",
    "    Scores residues during pairwise based on Bio.Align.substitution_matrices\n",
    "    Example: SubsMatCmp(substitution_matrices.load(\"BLOSUM62\")).cmp('A', 'W') == -3.0\n",
    "    \"\"\"\n",
    "    def __init__(self, matrix):\n",
    "        self.mx = matrix\n",
    "\n",
    "    def cmp(self, x, y) -> int:\n",
    "        return int(self.mx[x.upper()][y.upper()])\n",
    "\n",
    "\n",
    "def fill_indices(record) -> int:\n",
    "    \"\"\"\n",
    "    For every residue fills letter annotation with original sequence index\n",
    "    :param record: SeqRecord, with or without gaps\n",
    "    :return: number of last residue\n",
    "    \"\"\"\n",
    "    # should not be printed into file and just store indices\n",
    "    record.letter_annotations['fbb_indices'] = [0] * len(record.seq)\n",
    "    original_index = 0\n",
    "    for seq_index in range(len(record.seq)):\n",
    "        if record.seq[seq_index] == '-':\n",
    "            continue\n",
    "        original_index += 1\n",
    "        record.letter_annotations['fbb_indices'][seq_index] = original_index\n",
    "    return original_index\n",
    "\n",
    "\n",
    "def fill_maparray(msa1, msa2, row) -> int:\n",
    "    \"\"\"\n",
    "    Fills veralign maparray into letter_notations.\n",
    "    Maparray contains columns indexes of same alpha-C number\n",
    "    :return: number of found matched residues\n",
    "    \"\"\"\n",
    "    matches = 0\n",
    "    f, s = 0, 0\n",
    "    msa1[row].letter_annotations['fbb_maparray'] = [0] * len(msa1[row])\n",
    "    msa2[row].letter_annotations['fbb_maparray'] = [0] * len(msa2[row])\n",
    "    while f <= len(msa1[row])-1 and s <= len(msa2[row])-1:\n",
    "        res1 = msa1[row].letter_annotations['fbb_indices'][f]\n",
    "        res2 = msa2[row].letter_annotations['fbb_indices'][s]\n",
    "        if res1 != 0 and res1 == res2:\n",
    "            matches += 1\n",
    "            msa1[row].letter_annotations['fbb_maparray'][f] = s+1\n",
    "            msa2[row].letter_annotations['fbb_maparray'][s] = f+1\n",
    "        if f == len(msa1[row])-1 or s == len(msa2[row])-1:\n",
    "            break\n",
    "        if res1 == 0:\n",
    "            f += 1\n",
    "        if res2 == 0:\n",
    "            s += 1\n",
    "        if res1 != 0 and res2 != 0:\n",
    "            f += 1\n",
    "            s += 1\n",
    "    return matches\n",
    "\n",
    "\n",
    "def fill_test_with_truth(test_record, truth_record):\n",
    "    \"\"\"\n",
    "    Fills test_rec with truth_rec into letter_annotations\n",
    "    :return: mutated test_rec\n",
    "    \"\"\"\n",
    "    if len(truth_record.seq) >= len(test_record.seq):\n",
    "        test_record.letter_annotations['FBB_TR'] = str(truth_record.seq)[:len(test_record.seq)]\n",
    "    elif len(truth_record.seq) < len(test_record.seq):\n",
    "        residue = len(test_record.seq) - len(truth_record.seq)\n",
    "        seq = str(truth_record.seq)\n",
    "        seq += '_' * residue\n",
    "        test_record.letter_annotations['FBB_TR'] = seq\n",
    "    return test_record\n",
    "\n",
    "\n",
    "def column_wise(\n",
    "        test_msa,\n",
    "        truth_msa,\n",
    "        col_annotation: str,\n",
    "        aa_cmp,\n",
    "        msa_annotation_prefix: str = '',\n",
    "        scorech=score_chr,\n",
    "    ):\n",
    "    \"\"\"\n",
    "    Compares corresponding columns with each other and writes bool score as column_annotations to test_msa.\n",
    "    :param test_msa: msa where to save scores\n",
    "    :param truth_msa: msa against which score will be calculated\n",
    "    :param col_annotation: name of annotation to save scores\n",
    "    :param msa_annotation_prefix: if given then writes count and percentage of scores (which tests as True) to annotations of test_sma\n",
    "    :param aa_cmp: comparator of residues, takes two aa and returns bool\n",
    "    :return: tuple of count and percentage of scores which tests as True\n",
    "    \"\"\"\n",
    "\n",
    "    assert len(test_msa) == len(truth_msa)\n",
    "\n",
    "    matches = 0\n",
    "    alnlen = test_msa.get_alignment_length()\n",
    "    test_msa.column_annotations[col_annotation] = [scorech(False)] * alnlen\n",
    "    for col in range(alnlen):\n",
    "        got = None\n",
    "        equals = True\n",
    "        for row in range(len(test_msa)):\n",
    "            tst = test_msa[row].seq[col].upper()\n",
    "            trh_col = test_msa[row].letter_annotations['fbb_maparray'][col]\n",
    "            if got is None:\n",
    "                got = trh_col\n",
    "            if got != trh_col:\n",
    "                equals = False\n",
    "                break\n",
    "            if trh_col == 0:\n",
    "                equals = False\n",
    "                break\n",
    "            \n",
    "            # substract 1 since maparray starts from 1\n",
    "            trh = truth_msa[row].seq[trh_col-1].upper()\n",
    "            if not aa_cmp(tst, trh):\n",
    "                equals = False\n",
    "                break\n",
    "        \n",
    "        if equals:\n",
    "            matches += 1\n",
    "        test_msa.column_annotations[col_annotation][col] = scorech(equals)\n",
    "\n",
    "    pcnt = matches * 100 / alnlen\n",
    "    test_msa.column_annotations[col_annotation] = ''.join(test_msa.column_annotations[col_annotation])\n",
    "    if len(msa_annotation_prefix) != 0:\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_cw_cnt'] = matches\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_cw_pcnt'] = pcnt\n",
    "    return matches, pcnt\n",
    "\n",
    "\n",
    "def pairwise(\n",
    "        test_msa,\n",
    "        truth_msa,\n",
    "        col_annotation: str,\n",
    "        aa_cmp,\n",
    "        threshold_index: str,\n",
    "        threshold = lambda t: t > 0,\n",
    "        msa_annotation_prefix: str= '',\n",
    "        scorech=score_chr,\n",
    "    ):\n",
    "    \"\"\"\n",
    "    Compares corresponding cells of corresponding columns with each other and writes scores as column_annotations to test_msa.\n",
    "    :param test_msa: msa where to save scores\n",
    "    :param truth_msa: msa against which score will be calculated\n",
    "    :param col_annotation: name of annotation to save scores\n",
    "    :param msa_annotation_prefix: if given then writes sum of scores (which tests as True) to annotations of test_sma\n",
    "    :param aa_cmp: comparator of residues, takes two aa and returns integer score for pair\n",
    "    :param threshold: function that takes column sum of pairs and returns True if threshold matched\n",
    "    :param threshold_index: where to write conservativeness\n",
    "    :return: tuple (sum of column scores, amount of pairs, amount of pairs with positive score)\n",
    "    \"\"\"\n",
    "\n",
    "    assert len(test_msa) == len(truth_msa)\n",
    "\n",
    "    sp = 0\n",
    "    pairs = 0\n",
    "    positive_pairs = 0\n",
    "\n",
    "    alnlen = test_msa.get_alignment_length()\n",
    "    test_msa.column_annotations[col_annotation] = [scorech(False)] * alnlen\n",
    "    test_msa.column_annotations[threshold_index] = [scorech(False)] * alnlen\n",
    "    for col in range(alnlen):\n",
    "        colsum = 0\n",
    "        got = None\n",
    "        equals = True\n",
    "        for row in range(len(test_msa)):\n",
    "            pairs += 1\n",
    "            tst = test_msa[row].seq[col].upper()\n",
    "            trh_col = test_msa[row].letter_annotations['fbb_maparray'][col]\n",
    "            if trh_col == 0:\n",
    "                continue\n",
    "            if got is None:\n",
    "                got = trh_col\n",
    "            if got != trh_col:\n",
    "                equals = False\n",
    "                break\n",
    "\n",
    "            \n",
    "            # substract 1 since maparray starts from 1\n",
    "            trh = truth_msa[row].seq[trh_col-1].upper()\n",
    "            score = int(aa_cmp(tst, trh))\n",
    "            if score > 0:\n",
    "                positive_pairs += 1\n",
    "            colsum += score\n",
    "\n",
    "        if not equals:\n",
    "            continue\n",
    "\n",
    "        sp += colsum\n",
    "        if threshold(colsum):\n",
    "            test_msa.column_annotations[threshold_index][col] = scorech(True)\n",
    "        if colsum >= 0:\n",
    "            test_msa.column_annotations[col_annotation][col] = scorech(colsum)\n",
    "\n",
    "    test_msa.column_annotations[col_annotation] = ''.join(test_msa.column_annotations[col_annotation])\n",
    "    test_msa.column_annotations[threshold_index] = ''.join(test_msa.column_annotations[threshold_index])\n",
    "    if len(msa_annotation_prefix) != 0:\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_pw_sum'] = sp\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_pw_amount'] = pairs\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_pw_matched'] = positive_pairs\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_sp_score'] = positive_pairs * 100 / pairs\n",
    "    return sp, pairs, positive_pairs\n",
    "\n",
    "\n",
    "def mark_case_by_conservateveness(msa, index_name: str):\n",
    "    \"\"\"\n",
    "    Changes cases of all sequences in msa corresponding to boolean index\n",
    "    \"\"\"\n",
    "    from Bio.Seq import Seq\n",
    "\n",
    "\n",
    "    for aln in msa:\n",
    "        seq = list(aln.seq)\n",
    "        for i in range(len(seq)):\n",
    "            if msa.column_annotations[index_name][i] != score_chr(True):\n",
    "                seq[i] = seq[i].lower()\n",
    "            else:\n",
    "                seq[i] = seq[i].upper()\n",
    "        aln.seq = Seq(''.join(seq))\n",
    "    return msa\n",
    "\n",
    "\n",
    "def find_conservative_ranges(msa, index_name: str):\n",
    "    \"\"\"\n",
    "    Finds all ranges of conservative columns.\n",
    "    :param index_name: name of boolean index which determines wheteher column is conservative\n",
    "    :return: list of tuples (start_test, end_test, start_truth, end_truth) of each conservative.\n",
    "    \"\"\"\n",
    "    # could decide column is conservative\n",
    "    # even if there is gapes.\n",
    "    # need to find first row without gap\n",
    "    # to find out position in truth alignment\n",
    "    def get_maparray(column):\n",
    "        for row in range(len(msa)):\n",
    "            maparray = msa[row].letter_annotations['fbb_maparray']\n",
    "            if len(index) > len(maparray) or maparray[column] == 0:\n",
    "                continue\n",
    "            return maparray[column]\n",
    "\n",
    "    ranges = []\n",
    "    index = msa.column_annotations[index_name]\n",
    "    flag = False\n",
    "    start = 0\n",
    "    for col in range(len(index)):\n",
    "        if not flag and index[col] == score_chr(True):\n",
    "            flag = True\n",
    "            start = col\n",
    "        elif flag and index[col] == score_chr(False):\n",
    "            flag = False\n",
    "            trh = get_maparray(col-1)\n",
    "            start_trh = get_maparray(start)\n",
    "            # should not be marked as conservative in any case\n",
    "            assert trh is not None\n",
    "            assert start_trh is not None\n",
    "            # maparray positions starts with 1 so no need to add\n",
    "            ranges.append((start+1, col, start_trh, trh))\n",
    "\n",
    "    return ranges\n",
    "\n",
    "\n",
    "def load_substitution_matrices() -> list:\n",
    "    \"\"\"\n",
    "    Returns list of tuples (matix_name, matrix) loaded from from Bio.Align.substitution_matrices.\n",
    "    Matrices having not all of residues will be filtered out\n",
    "    \"\"\"\n",
    "    from Bio.Align import substitution_matrices as sm\n",
    "\n",
    "\n",
    "    out = [(name, sm.load(name)) for name in sm.load()]\n",
    "    residues = set([r.upper() for r in aa_classes.keys()])\n",
    "    def check(item):\n",
    "        for aa in residues:\n",
    "            if aa not in item[1].alphabet and aa != 'J':\n",
    "                return False\n",
    "        return True\n",
    "\n",
    "    return sorted(filter(check, out), key=lambda x: x[0])\n",
    "\n",
    "\n",
    "def compare_alignments(query, subject, threshold: int = 0, mark: str = None, sub_mxs=[]):\n",
    "    \"\"\"\n",
    "    :return: tuple (mutated_query, conservative_ranges or None if mark index not given)\n",
    "    \"\"\"\n",
    "    query.sort(key=lambda x: x.id)\n",
    "    subject.sort(key=lambda x: x.id)\n",
    "\n",
    "    for record in query:\n",
    "        fill_indices(record)\n",
    "    for record in subject:\n",
    "        fill_indices(record)\n",
    "    for i in range(min(len(query), len(subject))):\n",
    "        fill_maparray(query, subject, i)\n",
    "        fill_test_with_truth(query[i], subject[i])\n",
    "\n",
    "    column_wise(query, subject, 'FBB_EQ', are_residues_eq, msa_annotation_prefix='FBB_EQ')\n",
    "    column_wise(query, subject, 'FBB_SM', are_residues_similar, msa_annotation_prefix='FBB_SM')\n",
    "\n",
    "    pairwise(\n",
    "        query,\n",
    "        subject,\n",
    "        'FBB_SP_EQ',\n",
    "        lambda x, y: int(are_residues_eq(x, y)),\n",
    "        'FBB_SP_EQ_TH',\n",
    "        msa_annotation_prefix='FBB_SP_EQ',\n",
    "    )\n",
    "    pairwise(\n",
    "        query,\n",
    "        subject,\n",
    "        'FBB_SP_SM',\n",
    "        lambda x, y: int(are_residues_similar(x, y)),\n",
    "        'FBB_SP_SM_TH',\n",
    "        lambda x: x >= int(threshold),\n",
    "        msa_annotation_prefix='FBB_SP_SM',\n",
    "    )\n",
    "\n",
    "    for sm in sub_mxs:\n",
    "        comparator = SubsMatCmp(sm[1])\n",
    "        pairwise(\n",
    "            query,\n",
    "            subject,\n",
    "            f'FBB_SP_{sm[0]}',\n",
    "            comparator.cmp,\n",
    "            f'FBB_SP_{sm[0]}_TH',\n",
    "            msa_annotation_prefix=f'FBB_SP_{sm[0]}',\n",
    "        )\n",
    "\n",
    "    if not isinstance(mark, str) or not len(mark) != 0:\n",
    "        return query, None\n",
    "\n",
    "    mark_case_by_conservateveness(query, mark)\n",
    "    ranges = find_conservative_ranges(query, mark)\n",
    "\n",
    "    return query, ranges\n",
    "\n",
    "\n",
    "def compare_alignments_files(\n",
    "        query=None,\n",
    "        subject=None,\n",
    "        output=None,\n",
    "        query_format=None,\n",
    "        subject_format=None,\n",
    "        output_stats=None,\n",
    "        output_matrices=None,\n",
    "        **kwargs,\n",
    "    ):\n",
    "    import io\n",
    "    from Bio import AlignIO\n",
    "    from Bio.AlignIO.StockholmIO import StockholmWriter\n",
    "\n",
    "\n",
    "    qry = AlignIO.read(query, query_format)\n",
    "    sbj = AlignIO.read(subject, subject_format)\n",
    "\n",
    "    if output_matrices and 'sub_mxs' not in kwargs:\n",
    "        kwargs['sub_mxs'] = load_substitution_matrices()\n",
    "    qry, ranges = compare_alignments(qry, sbj, **kwargs)\n",
    "    if ranges:\n",
    "        print(\"Position in query,Position in subject\")\n",
    "        for rng in ranges:\n",
    "            print(f'{rng[0]}-{rng[1]},{rng[2]}-{rng[3]}')\n",
    "\n",
    "    # biopython 1.83 stockholm writer lacks\n",
    "    # support of writing (not reading though)\n",
    "    # file-level GF annotatioins.\n",
    "    # biopython required that column_annotations\n",
    "    # aligned by length with alignment length,\n",
    "    # and letter_annotations aligned by length\n",
    "    # with length of annotated sequence.\n",
    "    # also, by format spec, file starts with version info,\n",
    "    # and ends with '//\\n'.\n",
    "    # so we end up with only option - catch all writes\n",
    "    # into buffer, insert stats into GF annotations\n",
    "    # between header (with version and etc) and sequences\n",
    "    # and only then write it to the file.\n",
    "    buf = io.StringIO()\n",
    "    writer = StockholmWriter(buf)\n",
    "    for record in qry:\n",
    "        for annotation_name in record.letter_annotations.keys():\n",
    "            if annotation_name.startswith('FBB_'):\n",
    "                writer.pfam_gr_mapping[annotation_name] = annotation_name\n",
    "    for annotation_name in qry.column_annotations.keys():\n",
    "        if annotation_name.startswith('FBB_'):\n",
    "            writer.pfam_gc_mapping[annotation_name] = annotation_name\n",
    "    writer.write_alignment(qry)\n",
    "\n",
    "    out = buf.getvalue()\n",
    "    if output_stats:\n",
    "        stats = []\n",
    "        for key, value in sorted(qry.annotations.items()):\n",
    "            stats.append(f'#=GF {key} {value}\\n')\n",
    "        stats = ''.join(stats)\n",
    "\n",
    "        view = buf.getvalue()\n",
    "        fst_nl = view.find('\\n')\n",
    "        snd_nl = view.find('\\n', fst_nl+1)\n",
    "        out = view[:snd_nl] + \"\\n\" + stats + view[snd_nl+1:]\n",
    "\n",
    "    with open(output, 'w') as f:\n",
    "        f.write(out)\n",
    "\n",
    "    return qry, sbj\n",
    "\n",
    "\n",
    "if __name__ == '__main__' and '__file__' in globals():\n",
    "    import argparse\n",
    "\n",
    "\n",
    "    parser = argparse.ArgumentParser(\n",
    "        description='This program compares two MSA and writes statistics to stockholm output.',\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--query\",\n",
    "        \"-r\",\n",
    "        required=True,\n",
    "        help=\"alignment to test against \\\"the truth\\\"\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--subject\",\n",
    "        \"-t\",\n",
    "        required=True,\n",
    "        help=\"alignment to compare with\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--query-format\",\n",
    "        \"-q\",\n",
    "        dest='query_format',\n",
    "        required=True,\n",
    "        help=\"format of query. required by Bio.AlignIO\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--subject-format\",\n",
    "        \"-s\",\n",
    "        dest='subject_format',\n",
    "        required=True,\n",
    "        help=\"format of subject. required by Bio.AlignIO\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--output\",\n",
    "        \"-o\",\n",
    "        required=True,\n",
    "        help=\"path to file to write output\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--threshold\",\n",
    "        \"-d\",\n",
    "        required=True,\n",
    "        help=\"threshold for pairwise conservative detection\"\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--mark-conservative-by-case\",\n",
    "        \"-m\",\n",
    "        dest='mark',\n",
    "        help=\"index name which will store sequences formated with lower case,\" +\n",
    "            \"if not conservative column, and upper case otherwise.\" +\n",
    "            \"stores nothing if not given\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--stats\",\n",
    "        \"-a\",\n",
    "        default=False,\n",
    "        action='store_true',\n",
    "        dest='output_stats',\n",
    "        help=\"whether should output statistics\"\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--matrices\",\n",
    "        \"-i\",\n",
    "        default=False,\n",
    "        action='store_true',\n",
    "        dest='output_matrices',\n",
    "        help=\"whether should build indices based on substitution matrices\"\n",
    "    )\n",
    "\n",
    "    args = parser.parse_args()\n",
    "    compare_alignments_files(**vars(args))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0a0fa49e-6112-4ec9-add2-a98969ea2def",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Position in query,Position in subject\n",
      "1-16,1-16\n",
      "38-45,40-47\n",
      "74-83,77-86\n",
      "91-93,94-96\n",
      "103-103,107-107\n",
      "116-124,124-132\n",
      "231-238,232-239\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(<<class 'Bio.Align.MultipleSeqAlignment'> instance (518 records of length 253) at 7fe0d8efbd30>,\n",
       " <<class 'Bio.Align.MultipleSeqAlignment'> instance (518 records of length 255) at 7fe0d976bee0>)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare_alignments_files(\n",
    "    query='reductase.sth',\n",
    "    subject='reductase.mafft.fa',\n",
    "    query_format='stockholm',\n",
    "    subject_format='fasta',\n",
    "    output='reductase.cmp.sth',\n",
    "    threshold=510,\n",
    "    mark='FBB_EQ',\n",
    "    output_stats=True,\n",
    "    output_matrices=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7482aa7f-2ea0-4cad-b92c-0a7fd9aff143",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "|cs score|sp score|\n",
       "|--|--|\n",
       "|21.74|54.46|\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython import display\n",
    "\n",
    "\n",
    "cs = !grep FBB_EQ_cw_pcnt reductase.cmp.sth\n",
    "sp = !grep FBB_SP_EQ_sp_score reductase.cmp.sth\n",
    "cs = float(cs[0].split(' ')[-1])\n",
    "sp = float(sp[0].split(' ')[-1])\n",
    "table=f'''|cs score|sp score|\n",
    "|--|--|\n",
    "|{cs:.2f}|{sp:.2f}|\n",
    "'''\n",
    "display.display_markdown(table, raw=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c84bd0e-89af-4524-a273-31dfb2ec6c40",
   "metadata": {},
   "source": [
    "## Интерпретируйте результат\n",
    "\n",
    "половина пар совпала (по sp), только пятая часть колонок полностью совпала (по cs).\n",
    "\n",
    "для выравнивания 518-ти последовательностей, это может говорить о том,\n",
    "\n",
    "что мафт выравнил хуже (судя по cs),\n",
    "\n",
    "хоть и не так значительно (судя по sp), как я ожидал бы для 518 последовательностей"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6f2fe23-2eb4-4d0d-afeb-4f616866c3d0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
