{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b149134a-2254-4157-a285-59d54dd09e07",
   "metadata": {},
   "source": [
    "# Задание 1\n",
    "\n",
    "## вопрос\n",
    "Выберите любой домен из базы данных Pfam. Единственное условие – в выравнивании seed должно быть хотя бы 20 последовательностей. Кратко опишите функцию домена. Заполните таблицу\n",
    "## ответ\n",
    "\n",
    "note: accession в stockholm файле указан вместе с `.26`, хотя в браузере выводится без него\n",
    "\n",
    "\n",
    "|     AC Pfam|ID Pfam|#SEED|#All|#SW|#architectures|#3D|Taxonomy|#eukaryota|#archaea|#bacteria|\n",
    "|------------|-------|-----|----|---|--------------|---|--------|----------|--------|---------|\n",
    "|PF01625(.26)|   PMSR|  518| 66k|327|           232| 30|     42k|      7771|    2050|    54948|\n",
    "\n",
    "\n",
    "Выбранный домен - Peptide methionine sulfoxide reductase\n",
    "\n",
    "Функция домена - This enzyme repairs damaged proteins. Methionine sulfoxide in proteins is reduced to methionine.\n",
    "\n",
    "# Задание 2\n",
    "\n",
    "## вопрос\n",
    "Скачайте выравнивание seed для Вашего домена и откройте его в\n",
    "JalView. Раскрасьте выравнивание по свойствам аминокислот и консервативности колонок. Найдите в выравнивании консервативные и функционально консервативные колонки. Найдите хотя бы один максимально достоверный блок (см. презентацию к семинару 11). Заполните таблицу\n",
    "## ответ\n",
    "\n",
    "ДЗ дано для практики использования JalView. Но, так как я зачем-то выбрал в качестве примера выравнивание размером 518, то находить консервативные колонки визуально очень сложно. Видно, например, что абсолютно консервативная колонка есть в единственном экземпляре. Но так же видно, что многие другие колонки почти полностью консервативны или функционально-консервативны. Необходимо найти колонки, где нарушений условий консервативности минимальны, но из 518 строк сделать эту задачу визуально оч муторно. Поэтому я напишу для этого свою программу и выведу в таблицу именно ее. Так как ДЗ все-таки по использованию JalView, то вывод программы будет использован для того, чтоб в JalView подобрать threshold так, чтобы получилось +- тоже самое, что в выводе моей программы, при этом не придется вглядываться в детали и пытаться глазами подсчитать сколько строк нарушают консервативность"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ae6b54de",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections.abc import Iterable\n",
    "\n",
    "\n",
    "aa_classes = {\n",
    "    'A': 'F',\n",
    "    'V': 'F',\n",
    "    'L': 'F',\n",
    "    'I': 'F',\n",
    "    'J': 'F',\n",
    "    'M': 'F',\n",
    "    'P': 'F',\n",
    "    'F': 'R',\n",
    "    'Y': 'R',\n",
    "    'W': 'R',\n",
    "    'G': 'Z',\n",
    "    'S': 'Z',\n",
    "    'T': 'Z',\n",
    "    'C': 'Z',\n",
    "    'Q': 'Z',\n",
    "    'N': 'Z',\n",
    "    'E': 'N',\n",
    "    'D': 'N',\n",
    "    'K': 'P',\n",
    "    'R': 'P',\n",
    "    'H': 'P',\n",
    "}\n",
    "\n",
    "\n",
    "def residues_from_class(classes: Iterable[str]) -> Iterable[str]:\n",
    "    cls = [cls.upper() for cls in classes]\n",
    "    return set([it[0] for it in filter(lambda it: it[1] in cls, aa_classes.items())])\n",
    "\n",
    "\n",
    "class Column:\n",
    "    def __init__(self, pos: int, rows: int) -> None:\n",
    "        self.pos = pos\n",
    "        self.rows = rows\n",
    "        self.residues = dict()\n",
    "        self.classes  = dict()\n",
    "\n",
    "    def __str__(self) -> str:\n",
    "        indices  = sum(self.residues.values())\n",
    "        residues = ''.join(self.residues.keys())\n",
    "        s = f'{self.pos + 1}:[{residues}]'\n",
    "        if indices == self.rows:\n",
    "            return s\n",
    "        return f'{s}[{indices}/{self.rows}]'\n",
    "\n",
    "    def add(self, residue: str) -> None:\n",
    "        residue = residue.upper()\n",
    "        self.residues[residue] = 1 + self.residues.get(residue, 0)\n",
    "        cls = aa_classes.get(residue)\n",
    "        if cls is not None:\n",
    "            self.classes[cls] = 1 + self.classes.get(cls, 0)\n",
    "\n",
    "    def remove(self, residues: Iterable[str]) -> int:\n",
    "        \"\"\"\n",
    "        Removes all residues from column\n",
    "        :param residues: set of residues to delete\n",
    "        :return: number of deleted residues\n",
    "        \"\"\"\n",
    "        deleted = 0\n",
    "        for res in residues:\n",
    "            if res not in self.residues.keys():\n",
    "                continue\n",
    "            deleted += 1\n",
    "            num = self.residues.pop(res)\n",
    "            cls = aa_classes.get(res)\n",
    "            if cls is None or cls not in self.classes.keys():\n",
    "                continue\n",
    "            self.classes[cls] -= num\n",
    "            if self.classes[cls] == 0:\n",
    "                self.classes.pop(cls)\n",
    "        return deleted\n",
    "\n",
    "\n",
    "def cons_to_md(alnrows: int, full: Iterable[Column], func: Iterable[Column], reliable: Iterable[Iterable[int]]) -> str:\n",
    "    import io\n",
    "\n",
    "\n",
    "    maxlen = max([len(str(col)) for col in full + func + ['seed', '100%-cons', 'func-cons', 'max-reliable']])\n",
    "    header_cell = '-' * maxlen\n",
    "    buf = io.StringIO()\n",
    "    buf.write(f\"|{'seed'.ljust(maxlen)}|{'100%-cons'.ljust(maxlen)}|{'func-cons'.ljust(maxlen)}|{'max-reliable'.ljust(maxlen)}|\\n\" + \\\n",
    "        \"|\" + '|'.join([header_cell]*4) + \"|\\n\",\n",
    "    )\n",
    "    rows = max(len(full), len(func), len(reliable))\n",
    "    for i in range(rows):\n",
    "        buf.write('|' + f'{alnrows if i == 0 else \"\"}'.ljust(maxlen) + '|')\n",
    "        for var in [full, func, reliable]:\n",
    "            buf.write(f'{var[i] if i < len(var) else \"\"}'.ljust(maxlen) + '|')\n",
    "        buf.write(f'\\n')\n",
    "\n",
    "    return buf.getvalue()\n",
    "\n",
    "\n",
    "def block_contains(\n",
    "        msa,\n",
    "        col_start: int,\n",
    "        col_end: int,\n",
    "        symbol: str = '-',\n",
    "        row_start: int = 0,\n",
    "        row_end: int = -1,\n",
    ") -> bool:\n",
    "    \"\"\"\n",
    "    Check whether block specified by coordinates contains symbol\n",
    "    :param row_end: if not given or given negative, then uses len(msa)\n",
    "    \"\"\"\n",
    "    if row_end < 0:\n",
    "        row_end = len(msa)\n",
    "    assert 0 <= col_start < col_end and 0 <= row_start < row_end <= len(msa)\n",
    "\n",
    "    for row in range(row_start, row_end):\n",
    "        seq = msa[row].seq\n",
    "        assert col_end <= len(seq)\n",
    "        if symbol in seq[col_start:col_end]:\n",
    "            return True\n",
    "\n",
    "    return False\n",
    "\n",
    "\n",
    "def find_conservatives(\n",
    "        msa,\n",
    "        full_threshold: int,\n",
    "        func_threshold: int,\n",
    "        mark: str = None,\n",
    "    ):\n",
    "    full = list()\n",
    "    func = list()\n",
    "    reliable = list()\n",
    "\n",
    "    alnlen = msa.get_alignment_length()\n",
    "    msa.column_annotations['FBB_FULL'] = ['-'] * alnlen\n",
    "    msa.column_annotations['FBB_FUNC'] = ['-'] * alnlen\n",
    "    msa.column_annotations['FBB_RLBL'] = ['-'] * alnlen\n",
    "\n",
    "    for col in range(len(msa[0].seq)):\n",
    "        column = Column(col, len(msa))\n",
    "\n",
    "        for row in range(len(msa)):\n",
    "            assert len(msa[row].seq) == len(msa[0].seq)\n",
    "            column.add(msa[row].seq[col])\n",
    "\n",
    "        if len(column.classes) == 0 or len(column.residues) == 0:\n",
    "            continue\n",
    "\n",
    "        if '-' in column.residues.keys():\n",
    "            continue\n",
    "\n",
    "        if len(column.residues) == 1:\n",
    "            full.append(column)\n",
    "            msa.column_annotations['FBB_FULL'][col] = '1'\n",
    "            continue\n",
    "\n",
    "        full_cons = list(filter(lambda x: x[1] >= full_threshold, column.residues.items()))\n",
    "        full_cons.sort(key=lambda x: -1 * x[1])\n",
    "        if len(full_cons) != 0:\n",
    "            column.remove(set(column.residues.keys()) - {full_cons[0][0]})\n",
    "            full.append(column)\n",
    "            msa.column_annotations['FBB_FULL'][col] = '1'\n",
    "            continue\n",
    "\n",
    "        if len(column.classes) == 1:\n",
    "            func.append(column)\n",
    "            msa.column_annotations['FBB_FUNC'][col] = '1'\n",
    "            continue\n",
    "\n",
    "        func_cons = list(filter(lambda x: x[1] >= func_threshold, column.classes.items()))\n",
    "        func_cons.sort(key=lambda x: -1 * x[1])\n",
    "        if len(func_cons) != 0:\n",
    "            column.remove(set(column.residues.keys()) - residues_from_class(func_cons[0][0]))\n",
    "            func.append(column)\n",
    "            msa.column_annotations['FBB_FUNC'][col] = '1'\n",
    "\n",
    "    msa.column_annotations['FBB_FULL'] = ''.join(msa.column_annotations['FBB_FULL'])\n",
    "    msa.column_annotations['FBB_FUNC'] = ''.join(msa.column_annotations['FBB_FUNC'])\n",
    "\n",
    "    if len(full) + len(func) < 2:\n",
    "        msa.column_annotations['FBB_RLBL'] = ''.join(msa.column_annotations['FBB_RLBL'])\n",
    "        if mark is not None and len(mark) != 0:\n",
    "            mark_case_by_conservateveness(msa, mark)\n",
    "        return msa, full, func, reliable\n",
    "\n",
    "    cols = sorted(full + func, key=lambda x: x.pos)\n",
    "    flag = False\n",
    "    start = 0\n",
    "    for i in range(1, len(cols)):\n",
    "        gaped = block_contains(msa, cols[i-1].pos, cols[i].pos + 1, '-')\n",
    "        if not flag and not gaped:\n",
    "            flag = True\n",
    "            start = cols[i-1].pos\n",
    "        elif flag and gaped:\n",
    "            flag = False\n",
    "            reliable.append((start + 1, cols[i-1].pos + 1))\n",
    "            for col in range(start, cols[i-1].pos + 1):\n",
    "                msa.column_annotations['FBB_RLBL'][col] = '1'\n",
    "\n",
    "    msa.column_annotations['FBB_RLBL'] = ''.join(msa.column_annotations['FBB_RLBL'])\n",
    "    if mark is not None and len(mark) != 0:\n",
    "        mark_case_by_conservateveness(msa, mark)\n",
    "    return msa, full, func, reliable\n",
    "\n",
    "\n",
    "def mark_case_by_conservateveness(msa, index_name: str):\n",
    "    \"\"\"\n",
    "    Changes cases of all sequences in msa corresponding to boolean index\n",
    "    \"\"\"\n",
    "    from Bio.Seq import Seq\n",
    "\n",
    "\n",
    "    for aln in msa:\n",
    "        seq = list(aln.seq)\n",
    "        for i in range(len(seq)):\n",
    "            if msa.column_annotations[index_name][i] != '1':\n",
    "                seq[i] = seq[i].lower()\n",
    "            else:\n",
    "                seq[i] = seq[i].upper()\n",
    "        aln.seq = Seq(''.join(seq))\n",
    "    return msa\n",
    "\n",
    "\n",
    "def write_to_sth(msa, output):\n",
    "    from Bio.AlignIO.StockholmIO import StockholmWriter\n",
    "\n",
    "\n",
    "    writer = StockholmWriter(output)\n",
    "    for annotation_name in msa.column_annotations.keys():\n",
    "        if annotation_name.startswith('FBB_'):\n",
    "            writer.pfam_gc_mapping[annotation_name] = annotation_name\n",
    "    writer.write_alignment(msa)\n",
    "\n",
    "\n",
    "def find_conservatives_from_file(\n",
    "        input_filepath: str = None,\n",
    "        input_format: str = None,\n",
    "        full_threshold: int = None,\n",
    "        func_threshold: int = None,\n",
    "        output: str = None,\n",
    "        mark: str = None,\n",
    "    ):\n",
    "    from Bio import AlignIO\n",
    "\n",
    "\n",
    "    msa, full, func, reliable = find_conservatives(\n",
    "        AlignIO.read(input_filepath, input_format),\n",
    "        int(full_threshold),\n",
    "        int(func_threshold),\n",
    "        mark,\n",
    "    )\n",
    "\n",
    "    if output is None or len(output) == 0:\n",
    "        return msa, full, func, reliable\n",
    "\n",
    "    with open(output, 'w') as f:\n",
    "        write_to_sth(msa, f)\n",
    "\n",
    "    return msa, full, func, reliable\n",
    "\n",
    "\n",
    "if __name__ == '__main__' and '__file__' in globals():\n",
    "    import argparse\n",
    "\n",
    "\n",
    "    parser = argparse.ArgumentParser(\n",
    "        description='This program searches for convservative, functionally convservative and maximally reliable columns of MSA.',\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--input\",\n",
    "        \"-i\",\n",
    "        required=True,\n",
    "        dest='input_filepath',\n",
    "        help=\"input filepath with alignment\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--input-format\",\n",
    "        \"-f\",\n",
    "        required=True,\n",
    "        dest='input_format',\n",
    "        help=\"format of query. required by Bio.AlignIO\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--full-threshold\",\n",
    "        \"-l\",\n",
    "        required=True,\n",
    "        dest='full_threshold',\n",
    "        help=\"threshold for detecting fully conservative column\"\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--func-threshold\",\n",
    "        \"-n\",\n",
    "        required=True,\n",
    "        dest='func_threshold',\n",
    "        help=\"threshold for detecting fully conservative column\"\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--output\",\n",
    "        \"-o\",\n",
    "        default='',\n",
    "        help=\"filepath to write down stockholm output. print markdown table to stdout, if not given\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--mark-conservative-by-case\",\n",
    "        \"-m\",\n",
    "        dest='mark',\n",
    "        default='',\n",
    "        help=\"index name which will store sequences formated with lower case,\" +\n",
    "            \"if not conservative column, and upper case otherwise.\" +\n",
    "            \"stores nothing if not given\",\n",
    "    )\n",
    "\n",
    "    args = parser.parse_args()\n",
    "    msa, full, func, reliable = find_conservatives_from_file(**vars(args))\n",
    "    if args.output == '':\n",
    "        print(cons_to_md(len(msa), full, func, reliable))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9a257eb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "|seed               |100%-cons          |func-cons          |max-reliable       |\n",
       "|-------------------|-------------------|-------------------|-------------------|\n",
       "|518                |7:[G][515/518]     |8:[CGS]            |(7, 10)            |\n",
       "|                   |10:[W][517/518]    |46:[GCS][515/518]  |(76, 80)           |\n",
       "|                   |76:[E][517/518]    |80:[IVLMA][517/518]|                   |\n",
       "|                   |119:[Y][515/518]   |92:[LIVMA]         |                   |\n",
       "|                   |234:[H]            |                   |                   |\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython import display\n",
    "\n",
    "\n",
    "msa, full, func, reliable = find_conservatives_from_file('PF01625.seed.sth', 'stockholm', 515, 515)\n",
    "display.display_markdown(cons_to_md(len(msa), full, func, reliable), raw=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d031577b",
   "metadata": {},
   "source": [
    "По таблице видно, что бОльшая часть полностью консервативных колонок консервативны по не менее чем в 99% (не менее 515) организмах,  а так же одна полностью консервативная колонка. Схожая картина с функционально-консервативными колонками (программа не записывает полностью консервативные колонки в функционально консервативные). Всего найдено 9 колонок, из которых 3 попадают в 5-ти колоночный максимальный блок (первый). Возможно это можно будет использовать как маркер для будущих сравнений выравниваний"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0385b00",
   "metadata": {},
   "source": [
    "\n",
    "# Задание 3\n",
    "\n",
    "## вопрос\n",
    "Сравнение программ множественного выравнивания. Выровняйте последовательности из seed любыми двумя программами множественного выравнивания. Сравните полученные выравнивания с выравниванием seed. Для этого необходимо найти совпадающие блоки в двух выравниваниях.\n",
    "\n",
    "## ответ\n",
    "\n",
    "Программа для сравнений для сравнений будет реализована по аналогии с VerAlign. Но, для демонстрации принципа использования и построения `sp score`, будет так же добавлена возможность использовать другие эвристики, для определения консервативности VerAlign добавляет за совпадение остаток по 1 баллу, а так же отнимает за несовпадение разрывов. Но можно, например, для более чувствительного сравнения, использовать score из матриц замен, добавляя его к сумме колонки. В прямом виде это почти не будет иметь биологического смысла и сравнивать выравнивания так не выйдет. Но score колонок на основе матриц замен можно использовать, чтобы показать гибкость подхода Sum-of-Pairs. Похожим образом Sum-of-Pairs использует [AL2CO](https://doi.org/10.1093/bioinformatics/17.8.700). Та же AL2CO использует еще и Entropy-based показатель, который можно хранить и вычилять как `sp score`, переиспользуя бОльшую часть кода.  Для простоты отладки, программа будет записывать score каждой колонки в аннотацию стокгольм файла. Из-за ограничения формата, по которому аннотации колонок должны быть строго по одному символу на файл, придется хранить score в виде юникод символов. В таких же аннотациях можно хранить независимо друг от друга и все остальные индексы.  По одному из таких индексов-аннотаций (с посчитанными score'ами) можно привести остатки во всем в выравнивании в большой или малый регистр в зависимости от того, яв-ся ли они консервативными"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9a51bcc8-02c6-4ae3-9173-18500b4a1ce8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ascii_zero = 48\n",
    "ascii_last = 126\n",
    "utf_shift = 128\n",
    "\n",
    "\n",
    "def score_chr(score) -> str:\n",
    "    \"\"\"\n",
    "    Maps columns score to single character (stockholm annotations allows only single char)\n",
    "    :return: printable ascii or utf-8\n",
    "    :raises ValueError: if unexpected type, got non-single-char-str, got negative score or bigger than printable utf-8\n",
    "    \"\"\"\n",
    "    if score is True:\n",
    "        return '1'\n",
    "    if score is False or score is None:\n",
    "        return '-'\n",
    "    if isinstance(score, str):\n",
    "        assert len(score) == 1\n",
    "        return score\n",
    "    if isinstance(score, int) or isinstance(score, float):\n",
    "        assert 0 <= score <= 0x10ffff\n",
    "        score = int(score)\n",
    "        # add code of ascii zero\n",
    "        # to print ascii digits and letters\n",
    "        if score + ascii_zero < ascii_last:\n",
    "            score += ascii_zero\n",
    "        else:\n",
    "            score += utf_shift\n",
    "        return chr(score)\n",
    "\n",
    "    raise ValueError(f'unsupported type for stockholm score annotation: {score}')\n",
    "\n",
    "\n",
    "aa_classes = {\n",
    "    'A': 'F',\n",
    "    'V': 'F',\n",
    "    'L': 'F',\n",
    "    'I': 'F',\n",
    "    'J': 'F',\n",
    "    'M': 'F',\n",
    "    'P': 'F',\n",
    "    'F': 'R',\n",
    "    'Y': 'R',\n",
    "    'W': 'R',\n",
    "    'G': 'Z',\n",
    "    'S': 'Z',\n",
    "    'T': 'Z',\n",
    "    'C': 'Z',\n",
    "    'Q': 'Z',\n",
    "    'N': 'Z',\n",
    "    'E': 'N',\n",
    "    'D': 'N',\n",
    "    'K': 'P',\n",
    "    'R': 'P',\n",
    "    'H': 'P',\n",
    "}\n",
    "\n",
    "\n",
    "for residue, cls in list(aa_classes.items()):\n",
    "    aa_classes[residue.lower()] = cls\n",
    "\n",
    "\n",
    "def are_residues_eq(x, y) -> bool:\n",
    "    \"\"\"\n",
    "    :return: True if X and Y equals\n",
    "    \"\"\"\n",
    "    return x in aa_classes and x.upper() == y.upper()\n",
    "\n",
    "\n",
    "def are_residues_similar(x, y) -> bool:\n",
    "    \"\"\"\n",
    "    :return: True if X and Y have similar properties\n",
    "    \"\"\"\n",
    "    x = aa_classes.get(x)\n",
    "    y = aa_classes.get(y)\n",
    "    return x is not None and x == y\n",
    "\n",
    "\n",
    "class SubsMatCmp:\n",
    "    \"\"\"\n",
    "    Scores residues during pairwise based on Bio.Align.substitution_matrices\n",
    "    Example: SubsMatCmp(substitution_matrices.load(\"BLOSUM62\")).cmp('A', 'W') == -3.0\n",
    "    \"\"\"\n",
    "    def __init__(self, matrix):\n",
    "        self.mx = matrix\n",
    "\n",
    "    def cmp(self, x, y) -> int:\n",
    "        return int(self.mx[x.upper()][y.upper()])\n",
    "\n",
    "\n",
    "def fill_indices(record) -> int:\n",
    "    \"\"\"\n",
    "    For every residue fills letter annotation with original sequence index\n",
    "    :param record: SeqRecord, with or without gaps\n",
    "    :return: number of last residue\n",
    "    \"\"\"\n",
    "    # should not be printed into file and just store indices\n",
    "    record.letter_annotations['fbb_indices'] = [0] * len(record.seq)\n",
    "    original_index = 0\n",
    "    for seq_index in range(len(record.seq)):\n",
    "        if record.seq[seq_index] == '-':\n",
    "            continue\n",
    "        original_index += 1\n",
    "        record.letter_annotations['fbb_indices'][seq_index] = original_index\n",
    "    return original_index\n",
    "\n",
    "\n",
    "def fill_maparray(msa1, msa2, row) -> int:\n",
    "    \"\"\"\n",
    "    Fills veralign maparray into letter_notations.\n",
    "    Maparray contains columns indexes of same alpha-C number\n",
    "    :return: number of found matched residues\n",
    "    \"\"\"\n",
    "    matches = 0\n",
    "    f, s = 0, 0\n",
    "    msa1[row].letter_annotations['fbb_maparray'] = [0] * len(msa1[row])\n",
    "    msa2[row].letter_annotations['fbb_maparray'] = [0] * len(msa2[row])\n",
    "    while f <= len(msa1[row])-1 and s <= len(msa2[row])-1:\n",
    "        res1 = msa1[row].letter_annotations['fbb_indices'][f]\n",
    "        res2 = msa2[row].letter_annotations['fbb_indices'][s]\n",
    "        if res1 != 0 and res1 == res2:\n",
    "            matches += 1\n",
    "            msa1[row].letter_annotations['fbb_maparray'][f] = s+1\n",
    "            msa2[row].letter_annotations['fbb_maparray'][s] = f+1\n",
    "        if f == len(msa1[row])-1 or s == len(msa2[row])-1:\n",
    "            break\n",
    "        if res1 == 0:\n",
    "            f += 1\n",
    "        if res2 == 0:\n",
    "            s += 1\n",
    "        if res1 != 0 and res2 != 0:\n",
    "            f += 1\n",
    "            s += 1\n",
    "    return matches\n",
    "\n",
    "\n",
    "def fill_test_with_truth(test_record, truth_record):\n",
    "    \"\"\"\n",
    "    Fills test_rec with truth_rec into letter_annotations\n",
    "    :return: mutated test_rec\n",
    "    \"\"\"\n",
    "    if len(truth_record.seq) >= len(test_record.seq):\n",
    "        test_record.letter_annotations['FBB_TR'] = str(truth_record.seq)[:len(test_record.seq)]\n",
    "    elif len(truth_record.seq) < len(test_record.seq):\n",
    "        residue = len(test_record.seq) - len(truth_record.seq)\n",
    "        seq = str(truth_record.seq)\n",
    "        seq += '_' * residue\n",
    "        test_record.letter_annotations['FBB_TR'] = seq\n",
    "    return test_record\n",
    "\n",
    "\n",
    "def column_wise(\n",
    "        test_msa,\n",
    "        truth_msa,\n",
    "        col_annotation: str,\n",
    "        aa_cmp,\n",
    "        msa_annotation_prefix: str = '',\n",
    "        scorech=score_chr,\n",
    "    ):\n",
    "    \"\"\"\n",
    "    Compares corresponding columns with each other and writes bool score as column_annotations to test_msa.\n",
    "    :param test_msa: msa where to save scores\n",
    "    :param truth_msa: msa against which score will be calculated\n",
    "    :param col_annotation: name of annotation to save scores\n",
    "    :param msa_annotation_prefix: if given then writes count and percentage of scores (which tests as True) to annotations of test_sma\n",
    "    :param aa_cmp: comparator of residues, takes two aa and returns bool\n",
    "    :return: tuple of count and percentage of scores which tests as True\n",
    "    \"\"\"\n",
    "\n",
    "    assert len(test_msa) == len(truth_msa)\n",
    "\n",
    "    matches = 0\n",
    "    alnlen = test_msa.get_alignment_length()\n",
    "    test_msa.column_annotations[col_annotation] = [scorech(False)] * alnlen\n",
    "    for col in range(alnlen):\n",
    "        got = None\n",
    "        equals = True\n",
    "        for row in range(len(test_msa)):\n",
    "            tst = test_msa[row].seq[col].upper()\n",
    "            trh_col = test_msa[row].letter_annotations['fbb_maparray'][col]\n",
    "            if got is None:\n",
    "                got = trh_col\n",
    "            if got != trh_col:\n",
    "                equals = False\n",
    "                break\n",
    "            if trh_col == 0:\n",
    "                equals = False\n",
    "                break\n",
    "            \n",
    "            # substract 1 since maparray starts from 1\n",
    "            trh = truth_msa[row].seq[trh_col-1].upper()\n",
    "            if not aa_cmp(tst, trh):\n",
    "                equals = False\n",
    "                break\n",
    "        \n",
    "        if equals:\n",
    "            matches += 1\n",
    "        test_msa.column_annotations[col_annotation][col] = scorech(equals)\n",
    "\n",
    "    pcnt = matches * 100 / alnlen\n",
    "    test_msa.column_annotations[col_annotation] = ''.join(test_msa.column_annotations[col_annotation])\n",
    "    if len(msa_annotation_prefix) != 0:\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_cw_cnt'] = matches\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_cw_pcnt'] = pcnt\n",
    "    return matches, pcnt\n",
    "\n",
    "\n",
    "def pairwise(\n",
    "        test_msa,\n",
    "        truth_msa,\n",
    "        col_annotation: str,\n",
    "        aa_cmp,\n",
    "        threshold_index: str,\n",
    "        threshold = lambda t: t > 0,\n",
    "        msa_annotation_prefix: str= '',\n",
    "        scorech=score_chr,\n",
    "    ):\n",
    "    \"\"\"\n",
    "    Compares corresponding cells of corresponding columns with each other and writes scores as column_annotations to test_msa.\n",
    "    :param test_msa: msa where to save scores\n",
    "    :param truth_msa: msa against which score will be calculated\n",
    "    :param col_annotation: name of annotation to save scores\n",
    "    :param msa_annotation_prefix: if given then writes sum of scores (which tests as True) to annotations of test_sma\n",
    "    :param aa_cmp: comparator of residues, takes two aa and returns integer score for pair\n",
    "    :param threshold: function that takes column sum of pairs and returns True if threshold matched\n",
    "    :param threshold_index: where to write conservativeness\n",
    "    :return: tuple (sum of column scores, amount of pairs, amount of pairs with positive score)\n",
    "    \"\"\"\n",
    "\n",
    "    assert len(test_msa) == len(truth_msa)\n",
    "\n",
    "    sp = 0\n",
    "    pairs = 0\n",
    "    positive_pairs = 0\n",
    "\n",
    "    alnlen = test_msa.get_alignment_length()\n",
    "    test_msa.column_annotations[col_annotation] = [scorech(False)] * alnlen\n",
    "    test_msa.column_annotations[threshold_index] = [scorech(False)] * alnlen\n",
    "    for col in range(alnlen):\n",
    "        colsum = 0\n",
    "        got = None\n",
    "        equals = True\n",
    "        for row in range(len(test_msa)):\n",
    "            pairs += 1\n",
    "            tst = test_msa[row].seq[col].upper()\n",
    "            trh_col = test_msa[row].letter_annotations['fbb_maparray'][col]\n",
    "            if trh_col == 0:\n",
    "                continue\n",
    "            if got is None:\n",
    "                got = trh_col\n",
    "            if got != trh_col:\n",
    "                equals = False\n",
    "                break\n",
    "\n",
    "            \n",
    "            # substract 1 since maparray starts from 1\n",
    "            trh = truth_msa[row].seq[trh_col-1].upper()\n",
    "            score = int(aa_cmp(tst, trh))\n",
    "            if score > 0:\n",
    "                positive_pairs += 1\n",
    "            colsum += score\n",
    "\n",
    "        if not equals:\n",
    "            continue\n",
    "\n",
    "        sp += colsum\n",
    "        if threshold(colsum):\n",
    "            test_msa.column_annotations[threshold_index][col] = scorech(True)\n",
    "        if colsum >= 0:\n",
    "            test_msa.column_annotations[col_annotation][col] = scorech(colsum)\n",
    "\n",
    "    test_msa.column_annotations[col_annotation] = ''.join(test_msa.column_annotations[col_annotation])\n",
    "    test_msa.column_annotations[threshold_index] = ''.join(test_msa.column_annotations[threshold_index])\n",
    "    if len(msa_annotation_prefix) != 0:\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_pw_sum'] = sp\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_pw_amount'] = pairs\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_pw_matched'] = positive_pairs\n",
    "        test_msa.annotations[f'{msa_annotation_prefix}_sp_score'] = positive_pairs * 100 / pairs\n",
    "    return sp, pairs, positive_pairs\n",
    "\n",
    "\n",
    "def mark_case_by_conservateveness(msa, index_name: str):\n",
    "    \"\"\"\n",
    "    Changes cases of all sequences in msa corresponding to boolean index\n",
    "    \"\"\"\n",
    "    from Bio.Seq import Seq\n",
    "\n",
    "\n",
    "    for aln in msa:\n",
    "        seq = list(aln.seq)\n",
    "        for i in range(len(seq)):\n",
    "            if msa.column_annotations[index_name][i] != score_chr(True):\n",
    "                seq[i] = seq[i].lower()\n",
    "            else:\n",
    "                seq[i] = seq[i].upper()\n",
    "        aln.seq = Seq(''.join(seq))\n",
    "    return msa\n",
    "\n",
    "\n",
    "def find_conservative_ranges(msa, index_name: str):\n",
    "    \"\"\"\n",
    "    Finds all ranges of conservative columns.\n",
    "    :param index_name: name of boolean index which determines wheteher column is conservative\n",
    "    :return: list of tuples (start_test, end_test, start_truth, end_truth) of each conservative.\n",
    "    \"\"\"\n",
    "    # could decide column is conservative\n",
    "    # even if there is gapes.\n",
    "    # need to find first row without gap\n",
    "    # to find out position in truth alignment\n",
    "    def get_maparray(column):\n",
    "        for row in range(len(msa)):\n",
    "            maparray = msa[row].letter_annotations['fbb_maparray']\n",
    "            if len(index) > len(maparray) or maparray[column] == 0:\n",
    "                continue\n",
    "            return maparray[column]\n",
    "\n",
    "    ranges = []\n",
    "    index = msa.column_annotations[index_name]\n",
    "    flag = False\n",
    "    start = 0\n",
    "    for col in range(len(index)):\n",
    "        if not flag and index[col] == score_chr(True):\n",
    "            flag = True\n",
    "            start = col\n",
    "        elif flag and index[col] == score_chr(False):\n",
    "            flag = False\n",
    "            trh = get_maparray(col-1)\n",
    "            start_trh = get_maparray(start)\n",
    "            # should not be marked as conservative in any case\n",
    "            assert trh is not None\n",
    "            assert start_trh is not None\n",
    "            # maparray positions starts with 1 so no need to add\n",
    "            ranges.append((start+1, col, start_trh, trh))\n",
    "\n",
    "    return ranges\n",
    "\n",
    "\n",
    "def load_substitution_matrices() -> list:\n",
    "    \"\"\"\n",
    "    Returns list of tuples (matix_name, matrix) loaded from from Bio.Align.substitution_matrices.\n",
    "    Matrices having not all of residues will be filtered out\n",
    "    \"\"\"\n",
    "    from Bio.Align import substitution_matrices as sm\n",
    "\n",
    "\n",
    "    out = [(name, sm.load(name)) for name in sm.load()]\n",
    "    residues = set([r.upper() for r in aa_classes.keys()])\n",
    "    def check(item):\n",
    "        for aa in residues:\n",
    "            if aa not in item[1].alphabet and aa != 'J':\n",
    "                return False\n",
    "        return True\n",
    "\n",
    "    return sorted(filter(check, out), key=lambda x: x[0])\n",
    "\n",
    "\n",
    "def compare_alignments(query, subject, threshold: int = 0, mark: str = None, sub_mxs=[]):\n",
    "    \"\"\"\n",
    "    :return: tuple (mutated_query, conservative_ranges or None if mark index not given)\n",
    "    \"\"\"\n",
    "    query.sort(key=lambda x: x.id)\n",
    "    subject.sort(key=lambda x: x.id)\n",
    "\n",
    "    for record in query:\n",
    "        fill_indices(record)\n",
    "    for record in subject:\n",
    "        fill_indices(record)\n",
    "    for i in range(min(len(query), len(subject))):\n",
    "        fill_maparray(query, subject, i)\n",
    "        fill_test_with_truth(query[i], subject[i])\n",
    "\n",
    "    column_wise(query, subject, 'FBB_EQ', are_residues_eq, msa_annotation_prefix='FBB_EQ')\n",
    "    column_wise(query, subject, 'FBB_SM', are_residues_similar, msa_annotation_prefix='FBB_SM')\n",
    "\n",
    "    pairwise(\n",
    "        query,\n",
    "        subject,\n",
    "        'FBB_SP_EQ',\n",
    "        lambda x, y: int(are_residues_eq(x, y)),\n",
    "        'FBB_SP_EQ_TH',\n",
    "        msa_annotation_prefix='FBB_SP_EQ',\n",
    "    )\n",
    "    pairwise(\n",
    "        query,\n",
    "        subject,\n",
    "        'FBB_SP_SM',\n",
    "        lambda x, y: int(are_residues_similar(x, y)),\n",
    "        'FBB_SP_SM_TH',\n",
    "        lambda x: x >= int(threshold),\n",
    "        msa_annotation_prefix='FBB_SP_SM',\n",
    "    )\n",
    "\n",
    "    for sm in sub_mxs:\n",
    "        comparator = SubsMatCmp(sm[1])\n",
    "        pairwise(\n",
    "            query,\n",
    "            subject,\n",
    "            f'FBB_SP_{sm[0]}',\n",
    "            comparator.cmp,\n",
    "            f'FBB_SP_{sm[0]}_TH',\n",
    "            msa_annotation_prefix=f'FBB_SP_{sm[0]}',\n",
    "        )\n",
    "\n",
    "    if not isinstance(mark, str) or not len(mark) != 0:\n",
    "        return query, None\n",
    "\n",
    "    mark_case_by_conservateveness(query, mark)\n",
    "    ranges = find_conservative_ranges(query, mark)\n",
    "\n",
    "    return query, ranges\n",
    "\n",
    "\n",
    "def compare_alignments_files(\n",
    "        query=None,\n",
    "        subject=None,\n",
    "        output=None,\n",
    "        query_format=None,\n",
    "        subject_format=None,\n",
    "        output_stats=None,\n",
    "        output_matrices=None,\n",
    "        **kwargs,\n",
    "    ):\n",
    "    import io\n",
    "    from Bio import AlignIO\n",
    "    from Bio.AlignIO.StockholmIO import StockholmWriter\n",
    "\n",
    "\n",
    "    qry = AlignIO.read(query, query_format)\n",
    "    sbj = AlignIO.read(subject, subject_format)\n",
    "\n",
    "    if output_matrices and 'sub_mxs' not in kwargs:\n",
    "        kwargs['sub_mxs'] = load_substitution_matrices()\n",
    "    qry, ranges = compare_alignments(qry, sbj, **kwargs)\n",
    "    if ranges:\n",
    "        print(\"Position in query,Position in subject\")\n",
    "        for rng in ranges:\n",
    "            print(f'{rng[0]}-{rng[1]},{rng[2]}-{rng[3]}')\n",
    "\n",
    "    # biopython 1.83 stockholm writer lacks\n",
    "    # support of writing (not reading though)\n",
    "    # file-level GF annotatioins.\n",
    "    # biopython required that column_annotations\n",
    "    # aligned by length with alignment length,\n",
    "    # and letter_annotations aligned by length\n",
    "    # with length of annotated sequence.\n",
    "    # also, by format spec, file starts with version info,\n",
    "    # and ends with '//\\n'.\n",
    "    # so we end up with only option - catch all writes\n",
    "    # into buffer, insert stats into GF annotations\n",
    "    # between header (with version and etc) and sequences\n",
    "    # and only then write it to the file.\n",
    "    buf = io.StringIO()\n",
    "    writer = StockholmWriter(buf)\n",
    "    for record in qry:\n",
    "        for annotation_name in record.letter_annotations.keys():\n",
    "            if annotation_name.startswith('FBB_'):\n",
    "                writer.pfam_gr_mapping[annotation_name] = annotation_name\n",
    "    for annotation_name in qry.column_annotations.keys():\n",
    "        if annotation_name.startswith('FBB_'):\n",
    "            writer.pfam_gc_mapping[annotation_name] = annotation_name\n",
    "    writer.write_alignment(qry)\n",
    "\n",
    "    out = buf.getvalue()\n",
    "    if output_stats:\n",
    "        stats = []\n",
    "        for key, value in sorted(qry.annotations.items()):\n",
    "            stats.append(f'#=GF {key} {value}\\n')\n",
    "        stats = ''.join(stats)\n",
    "\n",
    "        view = buf.getvalue()\n",
    "        fst_nl = view.find('\\n')\n",
    "        snd_nl = view.find('\\n', fst_nl+1)\n",
    "        out = view[:snd_nl] + \"\\n\" + stats + view[snd_nl+1:]\n",
    "\n",
    "    with open(output, 'w') as f:\n",
    "        f.write(out)\n",
    "\n",
    "    return qry, sbj\n",
    "\n",
    "\n",
    "if __name__ == '__main__' and '__file__' in globals():\n",
    "    import argparse\n",
    "\n",
    "\n",
    "    parser = argparse.ArgumentParser(\n",
    "        description='This program compares two MSA and writes statistics to stockholm output.',\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--query\",\n",
    "        \"-r\",\n",
    "        required=True,\n",
    "        help=\"alignment to test against \\\"the truth\\\"\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--subject\",\n",
    "        \"-t\",\n",
    "        required=True,\n",
    "        help=\"alignment to compare with\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--query-format\",\n",
    "        \"-q\",\n",
    "        dest='query_format',\n",
    "        required=True,\n",
    "        help=\"format of query. required by Bio.AlignIO\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--subject-format\",\n",
    "        \"-s\",\n",
    "        dest='subject_format',\n",
    "        required=True,\n",
    "        help=\"format of subject. required by Bio.AlignIO\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--output\",\n",
    "        \"-o\",\n",
    "        required=True,\n",
    "        help=\"path to file to write output\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--threshold\",\n",
    "        \"-d\",\n",
    "        required=True,\n",
    "        help=\"threshold for pairwise conservative detection\"\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--mark-conservative-by-case\",\n",
    "        \"-m\",\n",
    "        dest='mark',\n",
    "        help=\"index name which will store sequences formated with lower case,\" +\n",
    "            \"if not conservative column, and upper case otherwise.\" +\n",
    "            \"stores nothing if not given\",\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--stats\",\n",
    "        \"-a\",\n",
    "        default=False,\n",
    "        action='store_true',\n",
    "        dest='output_stats',\n",
    "        help=\"whether should output statistics\"\n",
    "    )\n",
    "    parser.add_argument(\n",
    "        \"--matrices\",\n",
    "        \"-i\",\n",
    "        default=False,\n",
    "        action='store_true',\n",
    "        dest='output_matrices',\n",
    "        help=\"whether should build indices based on substitution matrices\"\n",
    "    )\n",
    "\n",
    "    args = parser.parse_args()\n",
    "    compare_alignments_files(**vars(args))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "703480fb",
   "metadata": {},
   "source": [
    "## Сравним колонки\n",
    "Для начала воспользуемся первой программой для того, чтобы сделать первые прикидки. Посмотрим насколько в наших выравниваниях совпадают консервативные столбцы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3b14e94c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "|seed               |100%-cons          |func-cons          |max-reliable       |\n",
       "|-------------------|-------------------|-------------------|-------------------|\n",
       "|518                |7:[G][515/518]     |8:[CGS]            |(7, 10)            |\n",
       "|                   |10:[W][517/518]    |83:[IVLMA][517/518]|(79, 83)           |\n",
       "|                   |79:[E][517/518]    |95:[LIVMA]         |                   |\n",
       "|                   |127:[Y][515/518]   |                   |                   |\n",
       "|                   |235:[H]            |                   |                   |\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython import display\n",
    "\n",
    "\n",
    "msa, full, func, reliable = find_conservatives_from_file('reductase.mafft.fa', 'fasta', 515, 515)\n",
    "display.display_markdown(cons_to_md(len(msa), full, func, reliable), raw=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3174a436",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "|seed               |100%-cons          |func-cons          |max-reliable       |\n",
       "|-------------------|-------------------|-------------------|-------------------|\n",
       "|518                |7:[G][515/518]     |8:[CGS]            |(7, 10)            |\n",
       "|                   |10:[W][517/518]    |47:[GCS][515/518]  |(82, 86)           |\n",
       "|                   |82:[E][517/518]    |86:[IVLMA][517/518]|                   |\n",
       "|                   |126:[Y][515/518]   |98:[LIVMA]         |                   |\n",
       "|                   |228:[H]            |                   |                   |\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython import display\n",
    "\n",
    "\n",
    "msa, full, func, reliable = find_conservatives_from_file('reductase.muscle.fa', 'fasta', 515, 515)\n",
    "display.display_markdown(cons_to_md(len(msa), full, func, reliable), raw=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80ed95c7",
   "metadata": {},
   "source": [
    "Некоторые консервативные колонки совпадают между всеми выравниваниями. Визуально в JalView с теми же порогами (в 99%) так же приходит в голову такой вывод. Так же и по табличкам (по номерам колонок) и визуально видно, что некоторые колонки съехали, так как разные методы по-разному расставили разрывы"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90eaf0ce",
   "metadata": {},
   "source": [
    "## Сравнение программой\n",
    "\n",
    "### Mafft\n",
    "\n",
    "Для начала сравним мафт и сид. Воспользуемся так же программой, данной в задании. Посмотрим будет ли разница между программами сравнений"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0a0fa49e-6112-4ec9-add2-a98969ea2def",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Position in query,Position in subject\n",
      "1-16,1-16\n",
      "40-47,38-45\n",
      "77-86,74-83\n",
      "94-96,91-93\n",
      "107-107,103-103\n",
      "124-132,116-124\n",
      "232-239,231-238\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(<<class 'Bio.Align.MultipleSeqAlignment'> instance (518 records of length 255) at 7fddc6907700>,\n",
       " <<class 'Bio.Align.MultipleSeqAlignment'> instance (518 records of length 253) at 7fddb3dfbbe0>)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare_alignments_files(\n",
    "    query='reductase.mafft.fa',\n",
    "    subject='PF01625.seed.sth',\n",
    "    query_format='fasta',\n",
    "    subject_format='stockholm',\n",
    "    output='reductase.cmp.mafft.sth',\n",
    "    threshold=510,\n",
    "    mark='FBB_EQ',\n",
    "    output_stats=True,\n",
    "    output_matrices=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8f5bb10",
   "metadata": {},
   "source": [
    "Так как программа, данная в задании, принимает на вход только фасты - переформатируем наш файл с сидом"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2dd7bebc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "\n",
    "AlignIO.write(AlignIO.read('PF01625.seed.sth', 'stockholm'), 'PF01625.seed.fa', 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "944058c3",
   "metadata": {},
   "source": [
    "Теперь сравним программы сравнения"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2897551d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12\n",
      "8\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "/opt/share/conda/bin/python3 compare_alignments.py -a2 'PF01625.seed.fa' -a1 reductase.mafft.fa > mafft.truth.txt\n",
    "/opt/share/conda/bin/python3 msa.comparsion.py -t 'PF01625.seed.sth' -r reductase.mafft.fa -s stockholm -q fasta  -o out.sth -d 515 -m FBB_EQ > mafft.test.txt\n",
    "\n",
    "cat mafft.truth.txt | wc -l\n",
    "cat mafft.test.txt | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "03a13b9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1c1\n",
      "< Position in alignment1,Position in alignment2\n",
      "---\n",
      "> Position in query,Position in subject\n",
      "3d2\n",
      "< 32-33,25-26\n",
      "5,7c4\n",
      "< 56-56,52-52\n",
      "< 77-90,74-87\n",
      "< 92-92,89-89\n",
      "---\n",
      "> 77-86,74-83\n",
      "11d7\n",
      "< 230-230,229-229\n"
     ]
    }
   ],
   "source": [
    "!diff mafft.truth.txt mafft.test.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bff76c55",
   "metadata": {},
   "source": [
    "Ввидно, что по большей части выводы программ совпали. Несовпадения были, в основном, в случаях одиночных колонок, что при такой длине и ширине выравнивания, я считаю, не яв-ся критичным."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "089b2189",
   "metadata": {},
   "source": [
    "## Muscle\n",
    "\n",
    "Теперь тоже самое с muscle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "edd34bb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Position in query,Position in subject\n",
      "1-15,1-15\n",
      "39-47,38-46\n",
      "80-89,74-83\n",
      "97-99,91-93\n",
      "102-104,96-98\n",
      "121-132,114-125\n",
      "225-232,231-238\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(<<class 'Bio.Align.MultipleSeqAlignment'> instance (518 records of length 247) at 7fddc69075e0>,\n",
       " <<class 'Bio.Align.MultipleSeqAlignment'> instance (518 records of length 253) at 7fddb3b8d2e0>)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare_alignments_files(\n",
    "    query='reductase.muscle.fa',\n",
    "    subject='PF01625.seed.sth',\n",
    "    query_format='fasta',\n",
    "    subject_format='stockholm',\n",
    "    output='reductase.cmp.muscle.sth',\n",
    "    threshold=510,\n",
    "    mark='FBB_EQ',\n",
    "    output_stats=True,\n",
    "    output_matrices=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "99b39452",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11\n",
      "8\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "/opt/share/conda/bin/python3 compare_alignments.py -a2 'PF01625.seed.fa' -a1 reductase.muscle.fa > muscle.truth.txt\n",
    "/opt/share/conda/bin/python3 msa.comparsion.py -t 'PF01625.seed.sth' -r reductase.muscle.fa -s stockholm -q fasta  -o out.sth -d 515 -m FBB_EQ > muscle.test.txt\n",
    "\n",
    "cat muscle.truth.txt | wc -l\n",
    "cat muscle.test.txt | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8271f268",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1c1\n",
      "< Position in alignment1,Position in alignment2\n",
      "---\n",
      "> Position in query,Position in subject\n",
      "5d4\n",
      "< 93-93,87-87\n",
      "8d6\n",
      "< 115-118,108-111\n",
      "10d7\n",
      "< 222-223,228-229\n"
     ]
    }
   ],
   "source": [
    "!diff muscle.truth.txt muscle.test.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f06e3fa4",
   "metadata": {},
   "source": [
    "Как видно в случае muscle программы для сравнений так же выдали крайне похожий результат. Отличия в двух из трех случаев были снова в одиночных колонках. Возможно обе программы годятся для сравнения выравниваний"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c06c424b",
   "metadata": {},
   "source": [
    "## CS + SP\n",
    "\n",
    "Сейчас результаты сравнений записаны в соответсвующие файлы - `reductase.cmp.mafft.sth`, `reductase.cmp.muscle.sth`.\n",
    "Выведем сравнения в таблицу"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0deefe89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "|method|cs score|sp score|\n",
       "|--|--|--|\n",
       "|mafft|21.57|58.15|\n",
       "|muscle|25.10|62.46|\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython import display\n",
    "\n",
    "\n",
    "cs_mafft = !grep FBB_EQ_cw_pcnt reductase.cmp.mafft.sth\n",
    "sp_mafft = !grep FBB_SP_EQ_sp_score reductase.cmp.mafft.sth\n",
    "cs_muscle = !grep FBB_EQ_cw_pcnt reductase.cmp.muscle.sth\n",
    "sp_muscle = !grep FBB_SP_EQ_sp_score reductase.cmp.muscle.sth\n",
    "cs_mafft = float(cs_mafft[0].split(' ')[-1])\n",
    "sp_mafft = float(sp_mafft[0].split(' ')[-1])\n",
    "cs_muscle = float(cs_muscle[0].split(' ')[-1])\n",
    "sp_muscle = float(sp_muscle[0].split(' ')[-1])\n",
    "table=f'|method|cs score|sp score|\\n|--|--|--|\\n|mafft|{cs_mafft:.2f}|{sp_mafft:.2f}|\\n|muscle|{cs_muscle:.2f}|{sp_muscle:.2f}|\\n'\n",
    "display.display_markdown(table, raw=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab09393",
   "metadata": {},
   "source": [
    "## Интерпретация\n",
    "\n",
    "1. В обоих выравниваниях около 60% пар совпало. По колонкам в обоих выравниваниях совпало около четверти. \n",
    "2. Но, так как у muscle оба скора больше, можно уверенно говорить, что он справился лучше мафта\n",
    "3. Учитывая такие низкие скоры, можно так же утверждать, что в изначальном сиде выравнивание гораздо лучше обоих - и mafft и muscle "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
